dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -216,59 +220,102 @@ private void reshuffleChannelsIfNecessary() {
+                int currentIndex, LimitedChannel channel, Response response, OptionalInt next) {
             if (next.isPresent()) {
-                log.debug(
-                        "Received throwable, switching to next channel",
-                        SafeArg.of("currentIndex", currentIndex),
-                        UnsafeArg.of("current", channel),
-                        SafeArg.of("nextIndex", next.getAsInt()),
-                        throwable);
-            } else {
-                log.debug(
-                        "Received throwable, but already switched",
-                        SafeArg.of("currentIndex", currentIndex),
-                        UnsafeArg.of("current", channel),
-                        throwable);
+                nextNodeBecauseResponseCode.mark();
+            }
+            if (log.isDebugEnabled()) {
+                if (next.isPresent()) {
+                    log.debug(
+                            "Received error status code, switching to next channel",
+                            SafeArg.of("status", response.code()),
+                            SafeArg.of("currentIndex", currentIndex),
+                            UnsafeArg.of("current", channel),
+                            SafeArg.of("nextIndex", next.getAsInt()));
+                } else {
+                    log.debug(
+                            "Received error status code, but we've already switched",
+                            SafeArg.of("status", response.code()),
+                            SafeArg.of("currentIndex", currentIndex),
+                            UnsafeArg.of("current", channel));
+                }
+            }
+        }
+
+        private void receivedThrowable(
+                int currentIndex, LimitedChannel channel, Throwable throwable, OptionalInt next) {
+            if (next.isPresent()) {
+                nextNodeBecauseThrowable.mark();
+            }
+            if (log.isDebugEnabled()) {
+                if (next.isPresent()) {
+                    log.debug(
+                            "Received throwable, switching to next channel",
+                            SafeArg.of("currentIndex", currentIndex),
+                            UnsafeArg.of("current", channel),
+                            SafeArg.of("nextIndex", next.getAsInt()),
+                            throwable);
+                } else {
+                    log.debug(
+                            "Received throwable, but already switched",
+                            SafeArg.of("currentIndex", currentIndex),
+                            UnsafeArg.of("current", channel),
+                            throwable);
+                }
+            }
+        }
+
+        private void instrumentSuccessfulResponse(int currentIndex, NodeList nodeList) {
+            if (nodeList.size() < 10) {
+                // hard limit ensures we don't create unbounded tags
+                metrics.success(Integer.toString(currentIndex));
```

Comment:
I think we're missing a `@CheckReturnValue` on the generated utility class, I imagine we want to mark this meter.
Can we create an array of meters and run `successfulResponseMeters.get(currentIndex).mark()`?

Reply:
Done